change labels() method return type

pradishb opened the pull request:
fix #57

pradishb committed:
change labels() method return type

pradishb committed:
add backwards compatibility for self

Llandy3d commented on pytheus/metrics.py:
```diff
@@ -16,6 +27,10 @@
 from pytheus.registry import REGISTRY, Collector, Registry
 from pytheus.utils import InfFloat, MetricType
 
+# Backwards compatibility for python<3.11
+if TYPE_CHECKING:
+    from typing import Self
```
I would skip this check and import directly a backward compatible backport 🤔

Llandy3d commented on pytheus/metrics.py:
```diff
@@ -202,7 +217,7 @@ def _raise_if_cannot_observe(self) -> None:
             raise UnobservableMetricException
 
-    def labels(self, pytheus_labels_: Optional[Labels] = None, **kwargs: str) -> "_Metric":
+    def labels(self, pytheus_labels_: Optional[Labels] = None, **kwargs: str) -> "Self":
```
If the suggestion from this review gets applied, I would remove the quotes and use directly `Self`

Llandy3d reviewed (changes requested):
👋  hi, thanks a lot for catching this issue and the contribution!

Due to having to support older versions of python I think the best course of action would be to use the [typing_extensions](https://github.com/python/typing_extensions) library to make direct use of `Self`.

From my tests using `"Self"` would still mark it as `unknown` for pyright the returned object (although it fixes the error on the method).

Let me know if you want to work on it further or if you prefer for me to take over, regardless of choice I really appreciate you taking the time for raising and fixing this 🙌 

So to recap the course of action would be:
1. Add a dependency to typing_extensions
2. Import `Self` without the need for the `TYP_CHECKING` check
3. Replace `"Self"` -> `Self`

pradishb committed:
use typing_extension for backwards compatibility

pradishb commented:
I agree that using typing_extensions is the better solution. I have made the requested changes,

Llandy3d reviewed (approved):
Looks good! 🙌 
Thank you for your contribution 🙇

Llandy3d commented:
it actually doesn't seem to be enough as the return type now mismatch 😭

pradishb committed:
Make _MetricCollector generic type

_MetricCollector used collect metric of type _Metric. However it should only collect metrics of type Self@_Metric.

_collector in class _Metric is now initialized with type Self@_Metric.

Llandy3d commented on pytheus/metrics.py:
```diff
@@ -146,7 +163,7 @@ def __init__(
         default_labels: Optional[Labels] = None,
         registry: Optional[Registry] = REGISTRY,
-        collector: Optional[_MetricCollector] = None,
+        collector: Optional[_MetricCollector[Self]] = None,
```
this needs to be changed also for the `Histogram` and `Summary` class 🤔

Llandy3d reviewed (commented):
Looks good!

I think that for completeness we should also align the typing for the mentioned classes, what do you think ?
Beside that looks great 🙌

pradishb committed:
fix: add missing generic type to collector parameter
